internal/js/tc55/timers/timers.go

```diff
@@ -0,0 +1,346 @@
+	if t == nil {
+		return nil // everything was cleared
+	}
+
+	err := t.task()
+
+	if e.queue.length() > 0 {
+		e.setupTaskTimeout()
+	} else {
+		e.freeEventLoopIfPossible()
+	}
+
+	return err
+}
+
+func (e *timers) setupTaskTimeout() {
+	e.queue.stopTimer()
+	delay := -time.Since(e.timers[e.queue.first().id])
+	if e.taskQueue == nil {
+		e.taskQueue = taskqueue.New(e.vu.RegisterCallback)
+		e.setupTaskQueueCloserOnIterationEnd()
+	}
+	q := e.taskQueue
+	e.queue.head = time.AfterFunc(delay, func() {
+		q.Queue(e.runFirstTask)
+	})
+}
+
+func (e *timers) closeTaskQueue() {
+	// this only runs on the event loop
+	if e.taskQueueCh == nil {
+		return
+	}
+	ch := e.taskQueueCh
+	// so that we do not execute it twice
+	e.taskQueueCh = nil
+
+	select {
+	case ch <- struct{}{}:
+		// wait for this to happen so we don't need to hit the event loop again
+		// instead this just closes the queue
+		<-ch
+	case <-e.vu.Context().Done(): // still shortcircuit if the context is done as we might block otherwise
+	}
+}
+
+// logger is helper to get a logger either from the state or the initenv
+func (e *timers) logger() logrus.FieldLogger {
+	if state := e.vu.State(); state != nil {
+		return state.Logger
+	}
+	return e.vu.InitEnv().Logger
+}
+
+func (e *timers) setupTaskQueueCloserOnIterationEnd() {
+	ctx := e.vu.Context()
+	q := e.taskQueue
+	ch := make(chan struct{})
+	e.taskQueueCh = ch
+	go func() {
+		select { // wait for one of the two
+		case <-ctx.Done():
+			// lets report timers won't be executed and clean the fields for the next execution
+			// we need to do this on the event loop as we don't want to have a race
+			q.Queue(func() error {
+				logger := e.logger()
+				for _, timer := range e.queue.queue {
+					logger.Warnf("%s %d was stopped because the VU iteration was interrupted",
+						timer.name, timer.id)
+				}
+
+				// TODO: use `clear` when we only support go 1.21 and above
+				e.timers = make(map[uint64]time.Time)
+				e.queue.stopTimer()
+				e.queue = new(timerQueue)
+				e.taskQueue = nil
+				return nil
+			})
+			q.Close()
+		case <-ch:
+			e.timers = make(map[uint64]time.Time)
+			e.queue.stopTimer()
+			e.queue = new(timerQueue)
+			e.taskQueue = nil
+			q.Close()
+			close(ch)
+		}
+	}()
+}
+
+// this is just a small struct to keep the internals of a timer
+type timer struct {
+	id          uint64
+	nextTrigger time.Time
+	task        func() error
+	name        string
+}
+
+// this is just a list of timers that should be ordered once after the other
+// this mostly just has methods to work on the slice
+type timerQueue struct {
+	queue []*timer
+	head  *time.Timer
+}
+
+func (tq *timerQueue) add(t *timer) int {
+	var i int
+	// don't use range as we want to index to go over one if it needs to go to the end
+	for ; i < len(tq.queue); i++ {
+		if tq.queue[i].nextTrigger.After(t.nextTrigger) {
+			break
+		}
+	}
+
+	tq.queue = append(tq.queue, nil)
+	copy(tq.queue[i+1:], tq.queue[i:])
+	tq.queue[i] = t
+	return i
+}
```

Comment:
I find this easier to grok, but it's up to you 🙇 

```suggestion
func (tq *timerQueue) add(t *timer) int {
	i := slices.IndexFunc(tq.queue, func(tt *timer) bool {
		return tt.nextTrigger.After(t.nextTrigger)
	})
	if i < 0 {
		i = len(tq.queue)
	}
	tq.queue = slices.Insert(tq.queue, i, t)
	return i
}
```